component-samples/http-device-sample/src/main/java/org/fidoalliance/fdo/sample/Device.java

```diff
@@ -469,9 +490,34 @@ protected MessagingService getMessagingService(Composite request) {
         WebClient client = new WebClient(url, dr, to2Dispatcher);
         client.run();
       } catch (RuntimeException e) {
-        logger.info("Unable to contact Owner at " + url + ". " + e.getMessage());
+        if (isCausedBy(e, SocketException.class)) {
+          logger.info("Unable to contact Owner at " + url + ". " + e.getMessage());
+        } else {
+          lastFailure = e;
+          try (StringWriter s = new StringWriter(); PrintWriter w = new PrintWriter(s)) {
+            logger.info(w);
+            logger.error("Unable to onboard from owner at "
+                    + url + ". " + e.getMessage() + "\n" + s);
+          } catch (IOException e2) {
+            logger.info(e2.getMessage());
+          }
+        }
       }
     }
+
+    if (null != lastFailure) {
+      fail(lastFailure);
+    }
+  }
+
+  boolean isCausedBy(Throwable e, Class<? extends Throwable> t) {
+    if (t.isAssignableFrom(e.getClass())) {
+      return true;
+    } else if (null != e.getCause()) {
+      return isCausedBy(e.getCause(), t);
```

Comment:
Why a recursive call is made here?

Reply:
Because exception chains are recursive, and we want to detect if a given type of exception occurs at any point in the chain.  So we check the current exception (the last one in the chain), then if a parent (cause) exists, we check that one, and so on.